engine/runtime/src/test/scala/org/enso/compiler/test/semantic/TypeSignaturesTest.scala

```diff
@@ -0,0 +1,239 @@
+          case None =>
+            Some((sig, expr, "the expression does not have a resolution"))
+          case Some(resolution) =>
+            if (resolution.target.qualifiedName == n) {
+              None
+            } else {
+              Some(
+                (
+                  sig,
+                  expr,
+                  s"The resolution is ${resolution.target.qualifiedName}, but expected ${n}"
+                )
+              )
+            }
+        }
+      case (Fn(args, res), t: IR.Type.Function) =>
+        if (args.length != t.args.length) {
+          return Some((sig, expr, "arity does not match"))
+        }
+        args
+          .lazyZip(t.args)
+          .flatMap(findInequalityWitness)
+          .headOption
+          .orElse(findInequalityWitness(res, t.result))
+      case (Union(items), t: IR.Type.Set.Union) =>
+        if (items.length != t.operands.length) {
+          return Some((sig, expr, "number of items does not match"))
+        }
+        items.lazyZip(t.operands).flatMap(findInequalityWitness).headOption
+      case _ => Some((sig, expr, "constructors are incompatible"))
+    }
+
+    override def apply(left: IR.Expression): MatchResult = {
+      findInequalityWitness(sig, left) match {
+        case Some((s, t, r)) =>
+          MatchResult(
+            matches = false,
+            s"""
+               |${left.showCode()}
+               |($left)
+               |does not match
+               |$sig.
+               |Analysis:
+               |  sub-expression
+               |    ${t.showCode()}
+               |    ($t)
+               |  did not match fragment $s
+               |  because $r.
+               |
+               |""".stripMargin,
+            "The type matched the matcher, but it should not."
+          )
+        case _ => MatchResult(matches = true, "", "")
+      }
+    }
+  }
+}
+
+class TypeSignaturesTest
+    extends AnyWordSpecLike
+    with Matchers
+    with TypeMatchers {
+
+  private val ctx = new InterpreterContext()
+  private val langCtx = ctx.ctx
+    .getBindings(LanguageInfo.ID)
+    .invokeMember(MethodNames.TopScope.LEAK_CONTEXT)
+    .asHostObject[Context]()
+
+  private val Module = QualifiedName.fromString("Unnamed.Test")
+
+  langCtx.getPackageRepository.registerSyntheticPackage("my_pkg", "My_Lib")
+  langCtx.getTopScope.createModule(
+    QualifiedName.fromString("my_pkg.My_Lib.Util"),
+    null,
+    s"""
+       |type Util_1
+       |type Util_2
+       |
+       |type Util_Sum
+       |    type Util_Sum_1
+       |    type Util_Sum_2
+       |""".stripMargin
+  )
+
+  implicit private class PreprocessModule(code: String) {
+    def preprocessModule: IR.Module = {
+      val module = new runtime.Module(Module, null, code)
+      langCtx.getCompiler.run(module)
+      module.getIr
+    }
+  }
+
+  private def getSignature(
+    module: IR.Module,
+    methodName: String
+  ): IR.Expression = {
+    val m = module.bindings.find {
```

Comment:
If we wanted to expose signatures to Enso programmers. For example via `Meta.signature`, then the implementation would be the same as code in this method, right?